### PR TITLE
fix(core): cap off-policy Horde scan sequence length before scan hang

### PR DESCRIPTION
**Pull request by rama0x1**
## Summary

`alberta_framework/core/off_policy_horde.py`'s `run_off_policy_horde_learning_loop`
hands its caller-supplied step arrays (`observations`, `cumulants`,
`next_observations`, `rhos`, and optionally `discounts`) straight to
`jax.lax.scan` with **no validation whatsoever** — no bound on the leading
(step) axis, no type check, and no check that the arrays share a length. A
hostile or mistaken caller supplying a huge array forces JAX to trace/compile
a scan of that length, hanging the process well before any step executes.

This is the same hang class already fixed for other scan-driven array loops
in this repo (`core/horde.py` #2016, `core/sarsa.py` #1996,
`core/partner_policy_fusion.py` #1991, `core/learners.py` #1989,
`utils/nexting.py` #1992) — `off_policy_horde.py` was not covered by any of
those PRs, and unlike some sibling Horde variants (e.g.
`independent_demon_horde.py`, `stacked_horde.py`), it had **no int32-budget
guard either**: the scan was completely unguarded.

`run_off_policy_horde_learning_loop` is exported public API
(`alberta_framework.__init__`); `run_off_policy_horde_learning_loop_batched`
calls it internally and inherits the guard transitively, so it is not
directly touched.

## Fix

Adds `_require_off_policy_horde_sequence_length` (exact-type JAX-array +
shape gate, rejects length outside `[1, 10_000]` before any scan) and
`_require_off_policy_horde_matching_length` (checks `cumulants`,
`next_observations`, `rhos`, and — when explicitly supplied — `discounts`
share `observations`' leading length). The `10_000` ceiling matches the
documented learning-loop step ceiling already established elsewhere in
`core` (`horde._HORDE_SEQUENCE_MAX_STEPS`,
`learners._LEARNING_LOOP_MAX_STEPS`, `sarsa._SARSA_SEQUENCE_MAX_STEPS`).

## Scope

- `alberta_framework/core/off_policy_horde.py`
- `tests/test_off_policy_horde_learning_loop_scan_hang.py` (new)

## Verification

Commit under test: `dff98896e926b2a989510391f4680bfdf5664bbe`

```text
# On origin/main ec035f73 (pre-fix): no validation exists at all in
# run_off_policy_horde_learning_loop -- confirmed by reading the pre-fix
# source (jax.lax.scan is called unconditionally on line 1546 with the
# caller-supplied leading length, no _require_* call anywhere above it).

# At this head: oversized (10_001), origin-hang-class-sized (200_000), and
# length-mismatched inputs (cumulants/rhos/discounts) are rejected with a
# ValueError/TypeError, and a monkeypatch spy proves
# alberta_framework.core.off_policy_horde.jax.lax.scan is never invoked.
# The documented last-fit length (10_000) and ordinary lengths still run
# end-to-end, including with an explicit discounts array.

.venv/bin/python -m pytest tests/test_off_policy_horde_learning_loop_scan_hang.py -q -o addopts=""
# 16 tests collected, 16 passed

.venv/bin/python -m pytest tests/test_off_policy_horde.py tests/test_off_policy_horde_update_working_set.py -q -o addopts=""
# 33 passed -- no regression in the existing off-policy Horde suite

.venv/bin/python -m ruff check alberta_framework/core/off_policy_horde.py tests/test_off_policy_horde_learning_loop_scan_hang.py
# All checks passed!

.venv/bin/python -m ruff check .
# All checks passed! (full repo)

.venv/bin/python -m mypy alberta_framework/core/off_policy_horde.py
# Success: no issues found in 1 source file

.venv/bin/python -m mypy tests/test_off_policy_horde_learning_loop_scan_hang.py
# Success: no issues found in 1 source file

.venv/bin/python -m mypy
# Success: no issues found in 197 source files (full repo, strict)

.venv/bin/python -c "import alberta_framework"  # import surface stays intact
```

- [x] logs: local pytest/ruff/mypy output above; 16/16 new tests passed, 33 passed across the existing off-policy Horde suite, zero mypy errors full-repo (197 files)

- [x] domain-artifact: N/A - scan sequence-length hang guard, no IPMNIST/benchmark shard produced

- AI assistance: yes

- Model(s) used: anthropic/claude-sonnet-5

- Agent tooling: Claude Code

- Skill path: contribute-to-asi (local skill copy)

- Provenance status: self-reported

AI provider/model: Anthropic / claude-sonnet-5
Client / agent tooling: Claude Code
Attribution status: self-reported

---

### CODE REVIEW

**Comment by lalalune** [comment]
Superseded by merged PR #2060, which consolidates this resource boundary with shared scan/work budgets, combined-work checks, and the broader regression panel.
